src/coordinator/api/v1/handler/prometheus/native/read.go

```diff
@@ -110,38 +97,104 @@ func (h *PromReadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
-func (h *PromReadHandler) parseRequest(r *http.Request) (string, *handler.ParseError) {
-	targetQueries, ok := r.URL.Query()[targetQuery]
+func (h *PromReadHandler) read(reqCtx context.Context, w http.ResponseWriter, params models.RequestParams) ([]ts.Series, error) {
+	ctx, cancel := context.WithTimeout(reqCtx, params.Timeout)
+	defer cancel()
+
+	opts := &executor.EngineOptions{
+	}
+	// Detect clients closing connections
+	abortCh, _ := handler.CloseWatcher(ctx, w)
+	opts.AbortCh = abortCh
+
+	parser, err := promql.Parse(params.Target)
+	if err != nil {
+		return nil, err
+	}
+
+	// Results is closed by execute
+	results := make(chan executor.Query)
+	// Block series slices are sorted by start time
+	seriesMap := make(map[string][]block.Series)
+	go h.engine.ExecuteExpr(ctx, parser, opts, params, results)
+
+	for result := range results {
+		if result.Err != nil {
+			return nil, err
+		}
+
+		blocks := result.Result.Blocks()
+		// TODO: Stream blocks to client
+		for blk := range blocks {
+			iter := blk.SeriesIter()
+			for iter.Next() {
+				insertSeriesInMap(iter.Current(), seriesMap)
+			}
+
+			blk.Close()
+		}
+	}
+
+	return seriesMapToSeriesList(seriesMap)
+}
+
+func insertSeriesInMap(blockSeries block.Series, seriesMap map[string][]block.Series) {
+	seriesID := blockSeries.Meta.Name
+	blockList, ok := seriesMap[seriesID]
 	if !ok {
-		return "", handler.NewParseError(errNoTargetFound, http.StatusBadRequest)
+		seriesMap[seriesID] = make([]block.Series, 1)
+		seriesMap[seriesID][0] = blockSeries
+		return
 	}
 
-	// NB(braskin): currently, we only support one query at a time
-	if len(targetQueries) > 1 {
-		return "", handler.NewParseError(errBatchQuery, http.StatusBadRequest)
+	// Insert sorted by start time
+	for idx, s := range blockList {
+		if blockSeries.Meta.Bounds.Start.Before(s.Meta.Bounds.Start) {
```

Comment:
Added TODO, will switch to binary once our blocklist gets big. For now, I'm not sure if we would even want to do it this way in long term. Maybe we just send this back to client in chunks

Reply:
i'm trying not to add things to context apart from request tracing and deadlines. Will work on a better way

Reply:
The nice thing about having it on the context is that it makes keeping track of resources really easy since we can just `defer ctx.Close()` in the handlers to clear everything safely